Various fixes for stacked items

luaanon opened the pull request:
All inspired by [Lenient Stack Size](https://github.com/ZoeyTheEgoist/FabricMods/tree/main/LenientStackSize) which has a [CC0 1.0 Universal License](https://github.com/ZoeyTheEgoist/FabricMods/blob/main/LenientStackSize/LICENSE).

Game version: 1.18.1

MixinAnvilScreenHandler
-  fix using stacks of enchanted books, and repairing stacks of items

MixinDispenserBehavior
-  fix entire stacks of buckets being replaced by a single bucket

MixinHorseScreenHandler
-  fix entire stacks being damaged at a time

MixinServerPlayNetworkHandler
-  fix entire stack of book and quill's being replaced with a single written book when signed

MixinJukeboxBlockEntity
-  fix using stack of discs on jukebox resulting in dupes. player stack goes down by one, but the jukebox stores and drops the full stack

luaanon committed:
Add files via upload

MixinAnvilScreenHandler
 fix using stacks of enchanted books, and repairing stacks of items

MixinDispenserBehavior
 fix entire stacks of buckets being replaced by a single bucket

MixinHorseScreenHandler
 fix entire stacks being damaged at a time

MixinServerPlayNetworkHandler
 fix entire stack of book and quill's being replaced with a single written book when signed

MixinJukeboxBlockEntity
 fix using stack of discs on jukebox resulting in dupes. player stack goes down by one, but the jukebox stores and drops the full stack

luaanon committed:
Add files via upload

Apply new mixins

KrisCris committed:
fix wrong itemStack being tested in decreaseOne of MixinAnvilScreen

KrisCris commented:
Hi,

I am busy these days and unable to check if these mixins are working properly. Can you please test them?  Also, I can't launch the game with `MixinDispenserBehavior` enabled.
```
Caused by: java.lang.RuntimeException: Mixin transformation of net.minecraft.block.dispenser.DispenserBehavior$8 failed
Caused by: org.spongepowered.asm.mixin.transformer.throwables.MixinTransformerError: An unexpected critical error was encountered
Caused by: org.spongepowered.asm.mixin.throwables.MixinApplyError: Mixin [allstackable.mixins.json:MixinDispenserBehavior] from phase [DEFAULT] in config [allstackable.mixins.json] from mod [allstackable] FAILED during APPLY
Caused by: org.spongepowered.asm.mixin.injection.throwables.InvalidInjectionException: @Inject annotation on dispenseOne specifies a target class 'net/minecraft/block/dispenser/ItemDispenserBehavior', which is not supported [PREINJECT Applicator Phase -> allstackable.mixins.json:MixinDispenserBehavior -> Prepare Injections ->  -> handler$zgl000$dispenseOne(Lnet/minecraft/util/math/BlockPointer;Lnet/minecraft/item/ItemStack;Lorg/spongepowered/asm/mixin/injection/callback/CallbackInfoReturnable;)V -> Parse]
```

Thanks for the contribution!

luaanon commented:
Good catch on the MixinAnvilScreen. I copy/pasted the isModified at the last minute to match your mixin style, and obviously goofed. I'll look over that check in all the other mixins to be safe. 

To be honest, most of the isModified checks aren't strictly needed in these patches...they were designed to work the same as vanilla when the stack count is 1. But there's no harm in it either, and it does isolate the modifications better.

I did test everything before adding the isModified checks, and they were working as described. I have no problems retesting them again, especially in light of the MixinAnvilScreen oversight. I'll follow up with another post describing the tests when I'm done with them.

======================================

As for MixinDispenserBehavior, it launches with no problems on my system.
-  Minecraft 1.18.1
-  Fabric Loader 0.12.11
-  fabric-api-0.44.0+1.18
-  allstackable 1.5.1-1.18.1 (modified with these commits)

 
 Also loads with:
-  Minecraft 1.18.1
-  Fabric Loader 0.12.12
-  fabric-api-0.45.2+1.18
-  allstackable 1.5.1-1.18.1 (modified with these commits)

 
======================================

Targeting the anonymous inner class is annoying, and I wouldn't be surprised if it's more brittle than expected.

Assuming that the current targeting is ultimately unworkable, other hooking options include:
1.  DispenserBlock::registerBehavior at head, and replace the entire DispenserBehavior when appropriate
2.  DispenserBehavior::registerDefaults after storing itemDispenserBehavior3,  and replace it with a new DispenserBehavior before it's used in any registerBehavior calls.
3.  DispenserBehavior::registerDefaults at return, and replace the old behaviors with new ones after everything's been registered. Could either use a hardcoded list of items, or make an accessor for DispenserBlock::BEHAVIORS and replace the behavior for every item that uses the same behavior as Items.LAVA_BUCKET.

All of them are quite heavy handed compared to the current injection, but should be markedly less brittle. 2 would be the most elegant, IMO, but also the most brittle of the three. 3 would be the least brittle and easiest to target. 1 would be fairly elegant as well, except for having to determine what "when appropriate" means exactly. 1 would also be the most likely to either work with modded items, or break modded items, unless "when appropriate" means a hardcoded list.

Do you have any preferences? I lean towards 3 with the accessor.

luaanon commented:
Here are the test results for MixinAnvilScreenHandler.

Works as described/intended, but found new issues.
- Renaming stacked items that are normally unstackable is broken, but still works for single items and normally stackable items. Without the patch, it does work for normally unstackable items.
- Enchantment costs for stacked items is broken, but still works for single items.

Also found two minor cosmetic issues
- Unable to shift-click items to refill input slot 0, but works fine with input slot 1.
- Shift-clicking the output when combining a stack with a book does work, but produces multiple stacks of 1 in the inventory instead of a single stacked stack.

Details are in [MixinAnvilScreenHandler.txt](https://github.com/KrisCris/AllStackable/files/7851033/MixinAnvilScreenHandler.txt)

luaanon commented:
Here are the test results for MixinDispenserBehavior.

Works as described/intended.

One cosmetic issue that I knew about, but didn't address:
1. It may be nicer to try and insert an empty bucket into the dispensers inventory, so that it has a chance to stack with any existing empty buckets and only dispense it if it can't stack anywhere. This diverges from vanilla behavior though, so I felt like it wasn't my call to make.

Details are in [MixinDispenserBehavior.txt](https://github.com/KrisCris/AllStackable/files/7851190/MixinDispenserBehavior.txt)

luaanon commented:
Here are the test results for MixinHorseScreenHandler.

Works as intended.

One note:
Saddles don't take damage, so my patch description was wrong. This simply prevents a mount from having more than one saddle equipped.

Details are in [MixinHorseScreenHandler.txt](https://github.com/KrisCris/AllStackable/files/7851215/MixinHorseScreenHandler.txt)

luaanon commented:
Here are the test results for MixinServerPlayNetworkHandler.

It's broken. In the same manner as MixinAnvilScreen when the isModified guard was added.

With the isModified guard removed or fixed a la MixinAnvilScreen, it works as originally tested.

Looking at it with fresh eyes, it has another issue that needs fixing. It assumes that the minecraft:written_book getMaxCount is greater than or equal to the minecraft:writable_book getMaxCount. If that isn't the case, the setCount will set the stack size higher than it should be. Instead, I guess it should make additional stacks and ItemsHelper.insertNewItem them into the player's inventory...though that will need a new injection point.

Details are in [MixinServerPlayNetworkHandler.txt](https://github.com/KrisCris/AllStackable/files/7851359/MixinServerPlayNetworkHandler.txt)

luaanon commented:
Here are the test results for MixinJukeboxBlockEntity.

Works as described/intended.

Details are in [MixinJukeboxBlockEntity.txt](https://github.com/KrisCris/AllStackable/files/7851401/MixinJukeboxBlockEntity.txt)

luaanon committed:
Merge branch 'KrisCris:1.18' into luaanon-patch-1

luaanon committed:
Fix two issues with writable and written books

The isModified guard on fixSignedBookCount now checks the correct stack
Splits the stack of written books into multiple stacks when the used writable book stacks count is greater than the written books maximum stack count

luaanon committed:
Additional fixes for AnvilScreenHandler

Fixed renaming stacked items that are normally not stackable
Fixed enchantment costs for stacked items that are normally not stackable

KrisCris commented:
Huge thanks to your contribution! I will merge this PR. :)

Seems even `MixinDispenserBehavior` stops the game running in my IDEA, I don't know why , it does work fine after built into jar.

Btw, did you manually script these .txt files, which looks like a lot of works to me lol

luaanon commented:
You're quite welcome :D 

Wrote them by hand with good old copy/paste. Bit of extra work, but helps me to think through the edge cases.
